### PR TITLE
meson: make git executable optional

### PR DESCRIPTION
**Pull request by Mic92**
This is useful in downstream packaging from a tarball

### CODE REVIEW

**Comment by xarkes** [comment]
I think it's wrong because the parameters `git_rev_list`, `git_describe` and `git_rev_parse` are required.
cc @pelijah

**Comment by pelijah** [comment]
So do you want to build r2 from tarballs of r2 and capstone? Actually R2_GITTIP, R2_GITTAP and R2_VERSION_COMMIT are not required. But without them users will not be able to determine full version of used r2 if they want to create an issue on this repo.

File: meson.build [hunk 1]
```diff
@@ -26,28 +26,27 @@ if meson.is_subproject()
...
-gittip  = run_command(git_exe, '-C', repo, 'rev-parse', 'HEAD')
-if gittip.returncode() != 0
-  gittip = 'unknown'
-else
-  gittip = gittip.stdout().strip()
+version_commit = 0
+gittap = ''
+gittip = 'unknown'
+if git_exe.found()
+    # Get version_commit
```
Comment (pelijah) [inline meson.build]: 2 spaces

File: shlr/meson.build [hunk 2]
```diff
@@ -2,7 +2,11 @@ res = run_command(py3_exe, '-c', '__import__("sys").exit(__import__("os").path.e
 if res.returncode() == 0
   message('Cloning capstone')
   git_cmd = 'clone -b next --depth 10 https://github.com/aquynh/capstone.git capstone'
-  run_command(git_exe, git_cmd.split())
+  if git_exe.found()
```
Comment (pelijah) [inline shlr/meson.build]: use `if not git_exe.found()` without `else` branch.

File: shlr/meson.build [hunk 3]
```diff
@@ -2,7 +2,11 @@ res = run_command(py3_exe, '-c', '__import__("sys").exit(__import__("os").path.e
...
+    run_command(git_exe, git_cmd.split())
+  else
+    error('git is required to download shlr/capstone')
```
Comment (pelijah) [inline shlr/meson.build]: And what about tarball?

Comment (Mic92) [reply shlr/meson.build]: In the tarball `shlr/capstone` exists, so it should not reach this branch right?

File: shlr/meson.build (hunk 2 above)
Comment (Mic92) [reply shlr/meson.build]: Just to clarify if the capstone directory does not exists, I should silently discard this error?

**Comment by Mic92** [comment]
If you want to enforce them all parameters can be also supplied by the user instead of git.

**Comment by xarkes** [comment]
What I mean is that to build radare2, you need to have those information which are important to identify the current radare2 version when reporting bug or asking for help.
In my humble opinion, building radare2 without those gittap information is pointless.

**Comment by Mic92** [comment]
@xarkes well is your policy not anyway to report bugs only from the master branch? In that case users of pre-build packages have to build from source anyway.

**Comment by xarkes** [comment]
Indeed.

**Comment by Mic92** [comment]
It is now possible supply the same information without dependency on git.

File: shlr/meson.build (hunk 2 above)
Comment (pelijah) [reply shlr/meson.build]: If capstone folder doesn't exist and git is not found print error.

File: shlr/meson.build (hunk 3 above)
Comment (pelijah) [reply shlr/meson.build]: Who distributes this tarball?

File: meson.build [hunk 4]
```diff
@@ -26,28 +26,28 @@ if meson.is_subproject()
...
-
-# Get gittip
-gittip  = run_command(git_exe, '-C', repo, 'rev-parse', 'HEAD')
-if gittip.returncode() != 0
-  gittip = 'unknown'
+if git_exe.found()
+  # Get version_commit
+  git_rev_list = run_command(git_exe, '-C', repo, 'rev-list', '--all', '--count')
+  if git_rev_list.returncode() == 0
+    version_commit = git_rev_list.stdout().strip()
```
Comment (pelijah) [inline meson.build]: No default value. Will fail if there is no `.git` directory.

File: meson.build [hunk 5]
```diff
@@ -26,28 +26,28 @@ if meson.is_subproject()
...
+  endif
+  
+  # Get gittap
+  git_describe = run_command(git_exe, '-C', repo, 'describe', '--tags', '--match', '[0-9]*')
+  if git_describe.returncode() == 0
+    gittap = git_describe.stdout().strip()
```
Comment (pelijah) [inline meson.build]: ^

File: meson.build [hunk 6]
```diff
@@ -26,28 +26,28 @@ if meson.is_subproject()
...
+  endif
+  
+  # Get gittip
+  git_rev_parse = run_command(git_exe, '-C', repo, 'rev-parse', 'HEAD')
+  if git_rev_parse.returncode() == 0
+    gittip = git_rev_parse.stdout().strip()
```
Comment (pelijah) [inline meson.build]: ^

File: shlr/meson.build (hunk 3 above)
Comment (Mic92) [reply shlr/meson.build]: You brought up the tarball. So I assumed there is a source tarball bundling capstone.

File: meson.build (hunk 6 above)
Comment (Mic92) [reply meson.build]: fixed

File: meson.build (hunk 5 above)
Comment (Mic92) [reply meson.build]: fixed

File: meson.build (hunk 4 above)
Comment (Mic92) [reply meson.build]: fixed

File: meson.build [hunk 7]
```diff
@@ -1,7 +1,7 @@
 project('radare2', 'c', meson_version: '>=0.46.0')
 
 py3_exe = import('python3').find_python()
-git_exe = find_program('git')
+git_exe = find_program('git', required : false)
```
Comment (pelijah) [inline meson.build]: Remove space before `:`

File: shlr/meson.build [hunk 8]
```diff
@@ -2,7 +2,9 @@ res = run_command(py3_exe, '-c', '__import__("sys").exit(__import__("os").path.e
 if res.returncode() == 0
   message('Cloning capstone')
   git_cmd = 'clone -b next --depth 10 https://github.com/aquynh/capstone.git capstone'
-  run_command(git_exe, git_cmd.split())
+  if git_exe.found()
```
Comment (pelijah) [inline shlr/meson.build]:
```meson
if not git_exe.found()
  error('blah-blah')
endif
run_command(git_exe, git_cmd.split())
```

**Comment by radare** [comment]
@xarkes can we merge this?

**Pull request merged**